### PR TITLE
feat(sdk): Add default registry context

### PR DESCRIPTION
**Pull request by ckadner**
**Description of your changes:**

Currently the new template registry client only recognizes host URLs that contain `kfp.pkg.dev` and only allows for registry server URLs containing `https://artifactregistry.googleapis.com/v1/projects/{project_id}/locations/{location}/repositories/{repo_id}/`.

This PR introduces a `default` registry context that enables connecting to non-`artifactregistry` servers without having to provide a registry context JSON file when connecting to a registry host.

/CC @chongyouquan @IronPan @hilcj

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/registry/registry_client.py [hunk 1]
```diff
@@ -129,8 +129,7 @@ def _is_ar_host(self) -> bool:
         Returns:
             Whether the host is on Artifact Registry.
         """
-        # TODO: Handle multiple known hosts.
-        return self._known_host_key == 'kfp_pkg_dev'
+        return self._known_host_key in _KNOWN_HOSTS_REGEX.keys()
```
Comment (chongyouquan) [inline sdk/python/kfp/registry/registry_client.py:132]: _is_ar_host is specifically for checking if the host is Artifact Registry.. It should just check if `self._known_host_key == 'kfp_pkg_dev'`.

**Review by chongyouquan — CHANGES_REQUESTED** [review]

File: sdk/python/kfp/registry/context/default_pkg_dev.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+   "download_tag_url": "{host}/{package_name}/{tag}",
+   "get_package_url": "{host}/packages/{package_name}",
+   "list_packages_url": "{host}/packages",
+   "delete_package_url": "{host}/packages/{package_name}",
+   "get_tag_url": "{host}/packages/{package_name}/tags/{tag}",
+   "list_tags_url": "{host}/packages/{package_name}/tags",
+   "delete_tag_url": "{host}/packages/{package_name}/tags/{tag}",
+   "create_tag_url": "{host}/packages/{package_name}/tags?tagId={tag}",
+   "update_tag_url": "{host}/packages/{package_name}/tags/{tag}?updateMask=version",
+   "get_version_url": "{host}/packages/{package_name}/versions/{version}?view=FULL",
```
[the remark this answers is not in the document]
Comment (ckadner) [reply sdk/python/kfp/registry/context/default_pkg_dev.json]: Makes sense. I removed them.

File: sdk/python/kfp/registry/registry_client.py (hunk 1 above)
Comment (ckadner) [reply sdk/python/kfp/registry/registry_client.py]: Right. I reinstated the original code here.

**Comment by ckadner** [comment]
Thanks for the quick review @chongyouquan! I update the code to address your comments.

**Comment by ckadner** [comment]
@connor-mccarthy @chensun -- I assume that there is nothing that I could do about the failing `kubeflow-pipelines-samples-v2` test as it appears infrastructure related? Please let me know if my assumption is incorrect.

**Comment by connor-mccarthy** [comment]
@ckadner, thanks for reaching out about those tests. Yes, that's correct. These tests are unfortunately quite flaky and we have planned development to address it.

**Comment by ckadner** [comment]
Thanks @chongyouquan !

/assign @connor-mccarthy

**Comment by ckadner** [comment]
Hi @connor-mccarthy -- I assigned you based on this bot comment. Thanks for taking a look!

> To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign connor-mccarthy after the PR has been reviewed.
You can assign the PR to them by writing /assign @connor-mccarthy in a comment when ready.

File: sdk/python/kfp/registry/registry_client_test.py [hunk 3]
```diff
@@ -48,10 +48,10 @@ def test_is_ar_host_kfp(self):
         self.assertEqual(client._is_ar_host(), True)
 
     def test_is_ar_host_error(self):
```
Comment (connor-mccarthy) [inline sdk/python/kfp/registry/registry_client_test.py:50]: nit: looks like this test name is out of date (no longer tests error)

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
This looks great! Thank you, @ckadner and @chongyouquan!

Just one small nitpick, then let's merge 👍🏻

File: sdk/python/kfp/registry/registry_client_test.py (hunk 3 above)
Comment (ckadner) [reply sdk/python/kfp/registry/registry_client_test.py]: Good catch. I updated it.

**Comment by ckadner** [comment]
Thanks for you review @connor-mccarthy! I updated the test name.

**Comment by connor-mccarthy** [comment]
/lgtm
/approve

Thanks, @ckadner!

**Comment by ckadner** [comment]
Thank you @connor-mccarthy and @chongyouquan!

**Comment by connor-mccarthy** [comment]
Manually merging. These changes do not affect sample tests.

**Pull request merged**